### PR TITLE
Sdk 760 py mongo instrumentation exception

### PR DESCRIPTION
**Pull request by haddasbronfman**

### CODE REVIEW

File: cisco_telescope/instrumentations/pymongo/__init__.py [hunk 1]
```diff
@@ -77,4 +77,6 @@ class ObjectIDEncoder(json.JSONEncoder):
     def default(self, obj):
         if isinstance(obj, objectid.ObjectId):
             return str(obj)
+        if isinstance(obj, pandas.Timestamp):
```
Comment (nemoshlag) [inline cisco_telescope/instrumentations/pymongo/__init__.py:80]: did you check if that timestamp object is by pandas I think error is "Object of type timestamp" and I'm afraid pandas.Timestamp tyoe won't catch all relevant cases

File: cisco_telescope/instrumentations/pymongo/__init__.py [hunk 2]
```diff
@@ -77,4 +77,6 @@ class ObjectIDEncoder(json.JSONEncoder):
...
+            return str(obj)
```
Comment (nemoshlag) [inline cisco_telescope/instrumentations/pymongo/__init__.py:81]: Not sure that str is the right "encoding" here, since users might want to filter by this field with "><=" operators

File: tests/instrumentations/test_pymongo.py [hunk 3]
```diff
@@ -2,9 +2,13 @@
 
 import pymongo
 from pymongo.errors import DuplicateKeyError
-
+import json
+import pandas as pd
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:6]: pd can be connected to many types, perhaps it's best to leave it as pandas for this case (pager duty, page decoder etc). That if we even need pandas to begin with

File: tests/instrumentations/test_pymongo.py [hunk 4]
```diff
@@ -47,6 +51,13 @@ def test_otel_attributes_unharmed(self):
             span.attributes[SpanAttributes.DB_STATEMENT], "insert customers"
         )
 
+    def test_json_with_timestamp_parsing(self):
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:54]: Since this is only testing encoder I would test it agains more complexed json (say real data from Zola or something similar)

File: tests/instrumentations/test_pymongo.py [hunk 5]
```diff
@@ -47,6 +51,13 @@ def test_otel_attributes_unharmed(self):
...
+        tstamp = pd.Timestamp("2023-01-01T12")
+        json_str = json.dumps(
+            {"created_at": tstamp}, cls=ObjectIDEncoder, skipkeys=True
+        )
+        print(json_str)
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:59]: redundant print

File: tests/instrumentations/test_pymongo.py [hunk 6]
```diff
@@ -47,6 +51,13 @@ def test_otel_attributes_unharmed(self):
...
+
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:60]: not sure what is tested in this test, just to verify exception isn't being raised?

File: cisco_telescope/instrumentations/pymongo/__init__.py (hunk 2 above)
Comment (osherv) [inline cisco_telescope/instrumentations/pymongo/__init__.py:81]: We should consider a more generic default.

**Comment by haddasbronfman** [comment]
After consulting with Haim & Nimrod, we decided to try/catch the decoder result, and if it fails, use str(obj).

File: requirements.txt [hunk 7]
```diff
@@ -173,4 +173,4 @@ Werkzeug==2.2.0
 wheel==0.37.1
 wrapt==1.14.1
 yarl==1.7.2
-zipp==3.8.1
+zipp==3.8.1
```
Comment (osherv) [inline requirements.txt:176]: What is this used for?

File: tests/instrumentations/test_pymongo.py [hunk 8]
```diff
@@ -47,6 +51,14 @@ def test_otel_attributes_unharmed(self):
             span.attributes[SpanAttributes.DB_STATEMENT], "insert customers"
         )
 
+    def test_json_with_timestamp_parsing(self):
+        timestamp = 1577916000  # selected date: 1/2/2020 10:00:00
+        dt_object = datetime.fromtimestamp(timestamp)
+        json_str = json.dumps(
+            {"created_at": dt_object}, cls=ObjectIDEncoder, skipkeys=True
+        )
+        self.assertEqual(json_str, '''{"created_at": "2020-01-01 22:00:00"}''')  # fmt: skip
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:60]: consider having the '# fmt: skip' comment above the non lint line, I find that more readable but maybe it's my personal taste

File: tests/instrumentations/test_pymongo.py [hunk 9]
```diff
@@ -47,6 +51,14 @@ def test_otel_attributes_unharmed(self):
...
+
```
Comment (osherv) [inline tests/instrumentations/test_pymongo.py:61]: I would separate this to another test class (which could be in this module, maybe 'ObjectIdEncoderTest'), and test more common types (bool, int, string, dict) and also assert that log was written

File: tests/instrumentations/test_pymongo.py [hunk 10]
```diff
@@ -47,6 +51,14 @@ def test_otel_attributes_unharmed(self):
             span.attributes[SpanAttributes.DB_STATEMENT], "insert customers"
         )
 
+    def test_json_with_timestamp_parsing(self):
+        timestamp = 1577916000  # selected date: 1/2/2020 10:00:00
+        dt_object = datetime.fromtimestamp(timestamp)
+        json_str = json.dumps(
```
Comment (nemoshlag) [inline tests/instrumentations/test_pymongo.py:57]: Since we're only testing the decoder I think more elaborate json is required here (see ordinary json object decoded properly and maybe another edge case)

File: cisco_telescope/instrumentations/pymongo/__init__.py [hunk 11]
```diff
@@ -75,6 +74,10 @@ def _uninstrument(self, **kwargs) -> None:
 
 class ObjectIDEncoder(json.JSONEncoder):
     def default(self, obj):
-        if isinstance(obj, objectid.ObjectId):
+        try:
+            return json.JSONEncoder.default(self, obj)
+        except TypeError as e:
+            logging.info(
+                f"Could not decode object of type: {type(obj)}, return is as a string, {e}"
```
Comment (nemoshlag) [inline cisco_telescope/instrumentations/pymongo/__init__.py:81]: I think there is a typo, maybe "return it as a string?"

File: requirements.txt (hunk 7 above)
Comment (haddasbronfman) [reply requirements.txt:176]: I have no idea. I also don't understand why it appears like I did a change in that line (because I didn't) This line is there before my PR.

File: tests/instrumentations/test_pymongo.py (hunk 8 above)
Comment (haddasbronfman) [reply tests/instrumentations/test_pymongo.py:60]: according to their docs I think this specific notation works when it comes at the end of the line: _"It doesn’t reformat lines that end with # fmt: skip or blocks that start with # fmt: off and end with # fmt: on."_
(I also tried to put it above the line and it didn't work, "make pretty" did change that line).

File: cisco_telescope/instrumentations/pymongo/__init__.py (hunk 11 above)
Comment (haddasbronfman) [reply cisco_telescope/instrumentations/pymongo/__init__.py]: fixed. sorry for that.

File: tests/instrumentations/test_pymongo.py (hunk 9 above)
Comment (haddasbronfman) [reply tests/instrumentations/test_pymongo.py]: Done. I create test_decoder and added a test for JSON parsing with string, int, double, bool, null, timestamp and array.

File: tests/instrumentations/test_pymongo.py (hunk 10 above)
Comment (haddasbronfman) [reply tests/instrumentations/test_pymongo.py]: I Accepted Osher's suggestion to create another file. so I create test_decoder.py and added tests for JSON parsing of string, int, double, bool, null, timestamp and array.

**Review by nemoshlag — DISMISSED** [review]

**Review by nemoshlag — APPROVED** [review]
